### DISCUSSION TITLE
Multiple Python inheritance in Gem5

### DISCUSSION THREAD

**Discussion by steven918** [gem5/gem5#1836] [Q&A]
It seems gem5 doesn't support multiple inheritance. For example

`class QoSMEM(MemCtrl, QoSMemCtrl)`

Would return "TypeErrro: SimObjects do not support multiple inheritance". 

The reason I'm trying to have multiple inheritance is because there seems no direct way to instantiate QoSMemCtrl parameters from MemCtrl-derived class. For example,

`class QoSMEM(MemCtrl):`
`      def __init__(self):`
`           super().__init__()`
`           self.write_high_thresh_perc= "95"    # OK, since write_high_thresh_perc is in MemCtrl`
`           self.qos_policy = "fifo"   # error because qos_policy is in QoSMemCtrl, even though MemCtrl is derived from QoSMemCtrl`

How to assign parameters in such multiple level inheritance?

Thanks

**Comment by powerjg** [comment]
What is the error for `self.qos_policy = "fifo"`? That should work.

Maybe it expects an enum object instead of a string? (I haven't looked at the code)

**Reply by steven918** [reply]
The error message of that line when using `class QoSMEM(MemCtrl)` is:

TypeError: SimObject.__init__() takes 1 positional argument but 2 were given
Error setting param QoSMEM.qos_policy to fifo

I verified that a string like "fifo" works by inheriting directly from QoSMemCtrl, like this:
`class QoSMEM(QosMemCtrl):`
` .... `
`     self.qos_policy = "fifo"`

So that string part should be fine.
